[C++] Improve Allocator handling

llchan opened the pull request:
- Templatize Allocator on vector_downward, and make it own the allocator
instance so it can manage lifetimes.
- Templatize + rename FlatBufferBuilderT accordingly, and add a typedef
to FlatBufferBuilder so old code continues to work.
- Fix some issues with the release deleter
- More details in github issue #4311 

Feedback welcome, I expect to iterate and discuss a bit before this is merged in.

EDIT: this is no longer template-based. See latest comments/changes.

llchan commented:
I've updated this PR with an inheritance-based approach. The important idea behind this remains the same: that the vector should own the allocator and be able to move it into the deleter when the buffer is released. The implementation is a little more verbose with the `std::unique_ptr<Allocator>`s, but without templates I think it's necessary to store a pointer to an external allocator. It's probably possible to do it with raw pointers too if anyone cares, but seems safer to use `unique_ptr`.

llchan commented:
Does anyone have access to a windows box with MSVC 16 to help me debug some remaining issues? There's probably a simple fix, but using commit+push+appveyer is a terrible workflow and I am calling it quits for now :)

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
+#if (!defined(_MSC_VER) || _MSC_VER > 1600)
+typedef std::unique_ptr<uint8_t, BufferDeleter> unique_ptr_t;
+#else  // begin hack for MSVC 16
+class unique_ptr_t {
```
Not so sure about this. If we were to provide our own `unique_ptr` emulation, I'd stick that in its own header, `stl_emulation.h` or similar. Though maybe its simpler to just not have this available on old platforms.
In either case, it be good to make sure one single define governs this (we already have `FLATBUFFERS_CPP98_STL` as opposed to making this MSVC specific.

llchan replied:
Yeah I wasn't _really_ intending for this to get merged, I just have no way of compiling MSVC other than pushing to this branch, so you're just seeing my experiments to see what the heck is wrong with MSVC.

On a more general note, I agree that we could probably refactor some of the compiler config checks into a standalone header (stuff like the final/override macros).

aardappel replied:
Not necessarily the compiler config stuff.. but this emulation code is only for a small number of users so could be tucked in a corner.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
-    : reserved_((initial_size + sizeof(largest_scalar_t) - 1) &
+                           std::unique_ptr<Allocator> &&allocator)
+
+    : allocator_(std::forward<std::unique_ptr<Allocator>>(allocator)),
```
Here we now have a core class dependent on newer STL, this may be hard to ifdef out. Could consider reverting this back to naked pointers, and documenting clearly that we want to own this after it is passed in (at the `FlatBufferBuilder` level)

llchan replied:
Yeah, I considered doing that, but it seems fairly dangerous. Maybe we could do something like what gRPC does with their slice wrappers, where they require the user to specify the ownerhsip transfer explicitly. Mocked off the top of my head here:

```c++ 
// The unique_ptr based implementation would be enabled where available
#ifndef FLATBUFFERS_CPP98_STL
FlatBufferBuilder(uoffset_t, std::unique_ptr<Allocator>&);
#endif

// Raw pointer-based implementations would require the user to be explicit about
// how ownership is transferred
enum StealAllocator { STEAL_ALLOCATOR };
FlatBufferBuilder(uoffset_t, Allocator*, StealAllocator);

enum BorrowAllocator { BORROW_ALLOCATOR };
FlatBufferBuilder(uoffset_t, Allocator*, BorrowAllocator);
```

This hints at something else I was going to add to this allocator PR, which is the ability to use shared allocators that are explictly _not_ owned by the vector. The plan was to add a proxy object that relays calls to the shared allocator. The use case might be memory pools and such, and I want to make sure we can support things like that.

aardappel replied:
I think that's a bit much for something that's for very few users and dying out. A typedef to have either naked or unique would be sufficient I think.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -123,6 +125,15 @@
+    (!defined(__GNUC__) || (__GNUC__ * 100 + __GNUC_MINOR__ >= 404))
+  #define FLATBUFFERS_DELETE_FUNC(func) func = delete;
+#else
+  #define FLATBUFFERS_DELETE_FUNC(func) private: func;
```
Yep, this is just doing the private declare for older compilers/stds. I think erring on the side of making things too private is better than pushing up to public by accident (in case the macro is used in a private/protected section), so let's leave it as is.

aardappel replied:
ok

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
+  // Reallocate `new_size` bytes of memory, replacing the old region of size
+  // `old_size` at `p`. In contrast to a normal realloc, this grows downwards,
+  // and is intended specifcally for `vector_downward` use.
+  virtual uint8_t *reallocate_downward(
```
`reallocate_downward` is actually overridden in the gRPC allocator, which is why I added this method. By letting the allocator handle the memcpy, the allocator no longer needs to have up to 2 allocate regions outstanding, and can just do the temp work in the function scope while it's reallocating. If you compare that PR's old/new impl you'll see that it's simplified significantly.

llchan replied:
As for why we don't merge `Allocator` and DefaultAllocator`... I suppose that's just a personal preference to keep interfaces fairly abstract and implementation-less. If you have a strong preference I'm fine with either.

aardappel replied:
ok

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
+
+  // Release myself to the caller by creating a new
+  // `std::unique_ptr<Allocator>` move of myself and resetting my own state.
+  virtual std::unique_ptr<Allocator> release() = 0;
```
Noted, will make sure these are disabled by that flag.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
+  }
+
+  // must stay at end of public section, may change access to private
+  FLATBUFFERS_DELETE_FUNC(BufferDeleter(const BufferDeleter& other))
```
Good catch. I've been using clang-format and I am trying to match convention, prob just missed that line.

llchan commented:
@aardappel appreciate the feedback on this. I hope it's not too much noise seeing all my work-in-progress commits, but I think we are making good progress and I hope the end result will be a more flexible + correct allocator interface.

llchan commented:
Also in general, if we are committed to supporting c++98, can we add it to the CI configs? Last I saw from skimming issues it seemed like there still an open discussion and there hadn't been a final say about whether we want to support old standards.

Fwiw my vote is for c++11 and best-effort support for older standards (e.g. dropping the unique_ptr stuff for c++98), rather than coding everything to the oldest standard.

aardappel commented:
@llchan no problem! I appreciate your work on this. Let me know when I should review again.

Btw the flag is not about C++98 language support.. we do require a minimal amount of C++11 (whatever VS2010 supports). The the particular use case is Android, where game developers often still compile against STLPort (which has no C++11 library features, but is compiled with a more modern compiler). We currently only support this use-case for flatbuffers.h + generated code files, a lot of the other code in the FlatBuffers project already doesn't compile with STLPort. Can't wait for the day when this is not necessary anymore.

I agree we really need CI for this. I've been meaning to look into Android CI but haven't gotten around to it, as with so many things.

llchan commented:
Ah got it. So the compiler/std is not 98, it's just that the STL implementation might not support the newer language features.

With the latest commit, we are finally passing the MSVC 16 tests now. A few notable changes since you last reviewed:
- Since `unique_ptr_t` has a custom deleter that is stateful and unsafe to mix with anything outside of flatbuffers, I actually opted to make a `DetachedBuffer` class to serve that role of transfering a buffer outside of a `FlatBufferBuilder`/`vector_downward`. This means everything shares the same implementation without the compiler-specific overrides, and I think some of the CPP98_STL-deprecated sections are no longer necessary. This fixes both VS2010 and CPP98_STL builds, I believe.
- I also removed `std::unique_ptr<Allocator>` uses, and replaced them with raw pointers with a steal flag. This introduced a new `AllocatorProxy` stub, which is just a non-owning proxy to an allocator. 
- I add some deprecated typedefs and stubs (`unique_ptr_t`, `simple_allocator`, and an old `FlatBufferBuilder` ctor), so that users can transition a little more gradually if they would like.

llchan commented:
@aardappel see what you think of the latest state of the PR. I think I've addressed most of the above points in a way that's  friendly to older compilers/stls.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
+// user to safely transfer ownership of a buffer. This resembles a unique_ptr +
+// custom deleter in many ways, but we do it here explicitly for wider compiler
+// support.
+class DetachedBuffer {
```
I appreciate your effort in making this more widely available, but this is a lot of code we're adding for a nice feature. We already had the situation where people on older compilers simply didn't have the "detach" functionality available, so we're not losing anything by using `unique_ptr` and ifdeffing it. Also, I'd like the "cost" (extra dynamic memory allocations) of using allocators and detaching to be paid by those using these features, and not everyone else (the 99% :)

llchan replied:
This wasn't just for CPP98_STL, it also fixed MSVC 16, which appears to have a wonky unique_ptr implementation. If we're fine with dropping `release()` for that I don't mind using `unique_ptr`. However, for what it's worth, the deleter involved with the `unique_ptr` is effectively the same thing as this, so it wouldn't save _that_ much code. Plus it's not truly a `unique_ptr` anyways (you can't swap in another pointer and expect the allocator not to blow up).

aardappel replied:
We haven't had problems with `unique_ptr` on VS2010 (==1600) before (including the code you're replacing), so not sure what the problem is here?
We do run the full set of tests on VS2010, which needs this functionality, so we cannot define for CPP98_STL for VS2010.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
-  void clear() {
-    if (buf_ == nullptr)
-      buf_ = allocator_.allocate(reserved_);
+  void reset(size_t initial_size, Allocator *allocator, bool steal_allocator) {
```
I don't see where this function is used.

llchan replied:
It's used in the other gRPC-related PR. When a message is popped out of a builder, we want to reset the vector with an allocator so it remains valid. Otherwise as soon as we pop out the message, the vector/builder is a landmine with nullptrs inside of it. I can take this out and rework that if necessary.

aardappel replied:
Ah.. might be nice to add it in the GRPC PR then, so we can see how caller and callee interact. For example, I do not see from this code why we don't simply keep the existing allocator when we reset?

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -123,6 +125,31 @@
+
+#if (__cplusplus >= 201402L) && defined(__has_cpp_attribute)
+  #if __has_cpp_attribute(deprecated)
+    #define DEPRECATED(msg, func) [[deprecated(msg)]] func
```
Okay, I just thought it'd make upgrades easier if there was a compiler warning/error telling you what changed. But we can certainly axe things aggressively instead.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
+
+// simple_allocator is retained for backwards compatibility, but we should
+// eventually deprecate and remove it.
+class simple_allocator : public DefaultAllocator {
```
Same as above, I was hoping to make upgrades more gradual by giving a deprecation rather than removing it entirely, but I'm fine with just nixing it if we don't need that.

aardappel replied:
For things that are very unlikely to be used externally, I think that is preferable, yes.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +563,228 @@ struct String : public Vector<char> {
+};
+
+// AllocatorProxy references an externally-managed allocator
+class AllocatorProxy : public Allocator {
```
The proxy is a non-owning reference to an allocator. An example would be a memory pool allocator that should be shared across many buffer builders, which means that the vectors should not free that allocator. Seemed like a simpler solution than a refcounted solution, but maybe there's something in between. I'll see what alternatives there are.

aardappel replied:
If we don't own the allocator, I'd rather that just be denoted by a bool, and make the client responsible for knowing when it is safe to deallocate the allocator. I agree that proper reference counting is nicer and safer, but we're talking about a niche use of a niche feature here, so simplicity is preferable.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -712,15 +901,23 @@ FLATBUFFERS_FINAL_CLASS
+                             Allocator *allocator = nullptr,
+                             bool steal_allocator = false)
+    : buf_(initial_size,
+           allocator ? allocator : new DefaultAllocator(),
```
This was introduced when we decided we didn't want templates and went to inheritance, which means an extra level of indirection, and we can't store the allocator as a member. I can modify the default case to point to a static `DefaultAllocator` since it's stateless.

llchan replied:
Templates would really improve a lot of this as far as efficiency goes, and with the right typedefs it'd be invisible to users who don't define custom allocators... might be worth reconsidering :)

aardappel replied:
Yes, as static default would be a nice savings for most users.
I'm generally a fan of solving problems with templates over inheritance.. but again, trying to keep the code nicest for the default case. I know I'm being peculiar.

llchan commented:
Let me take a step back... what's the use case for `release()`? Are users intended to actually use the data pointer there? They don't have a size, so seems like the use case for it is pretty limited, unless I'm missing something. Do we actually just want move ctor/assignment for `FlatBufferBuilder` so they can move the buffer around?

aardappel commented:
The use case for `release()` is for people to be able to hang on to a buffer they just constructed without having to memcpy it elsewhere, and without having to hold on to a dynamically allocated FlatBufferBuilder (which contains other data not needed after construction).

llchan commented:
How would they use the buffer though, without an associated size? It seems that the `DetachedBuffer` class does what we want, and using `unique_ptr` for it is kind of incorrect, because it doesnt act like a normal `unique_ptr` (e.g. no reset to another pointer).

llchan commented:
Sorry couldnt work on this yesterday, just made some changes now. Some notes:
- Deprecations have been removed, and we are aggressively breaking the old `simple_allocator` API on purpose.
- There is now a static `DefaultAllocator` that everyone can reference without a dynamic memory allocation.
- There is now a new bool boolean flag indicating whether the builder/vector/buffer owns the allocator and should free it, and this is transferred as necessary. This eliminates the `AllocatorProxy` class and eliminates heap allocations for the common (i.e. non-owning) case.
- The more I look at it, the more I think we should be more precise about what `release()` or `ReleaseBufferPointer()` returns. For example, in the `test.cpp`, the tests get the pointer from the returned `flatbuf`, but since it doesnt have a size, it gets the size from a string from a different test. With this change, the returned `flatbuf` is self-contained and has enough information to safely use the buffer, and usage should be cleaner.
- The old `DetachedBuffer` is now called `FlatBuffer`. The rationale is that a `FlatBufferBuilder` builds the message, and when you call `Release()`, it gives you a finished/fixed `FlatBuffer` that now owns the memory.
- I've removed the `reset(...)` method that isn't used here locally. Will include with the gRPC PR as necessary.

llchan commented:
Also, I'd like for the `FlatBuffer` class to have a `Verify<T>()` and `GetRoot<T>()`, so that a released `FlatBuffer` can be used easily. However, this would require some fairly aggressive reordering in the header file, so I will defer that for later.

As a general question, is there a reason to put everything in one header, as opposed to smaller ones that `flatbuffers.h` includes? It generates a huge diff here if I move anything around, and we could avoid that if things were split up.

aardappel commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -712,15 +844,21 @@ FLATBUFFERS_FINAL_CLASS
-        dedup_vtables_(true), string_pool(nullptr) {
+                             Allocator *allocator = nullptr,
+                             bool own_allocator = false)
+    : buf_(initial_size, allocator, own_allocator),
```
As above, the Google C++ Style Guide actually wants to put multiple items per line :)

llchan replied:
Updated my clang-format rules for this. Not sure why you guys like that, the diffs are cleaner and it's easier to read with them split onto separate lines, but I'll follow project conventions :)

aardappel replied:
Not sure either. There's a lot in the style guide I disagree with. In this case I would say it is generally good to just keep the existing code, your diff becomes more complicated with "drive by" code style improvements that may just be a preference :)

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +547,186 @@ struct String : public Vector<char> {
+// FlatBuffer is a finished flatbuffer memory region, detached from its
+// builder. The original memory region and allocator are also stored so that
+// the FlatBuffer can manage the memory lifetime.
+class FlatBuffer {
```
Renamed back to `DetachedBuffer`

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +547,186 @@ struct String : public Vector<char> {
+    assert(allocator_);
+  }
+
+  FlatBuffer(FlatBuffer &&other)
```
What's the appropriate ifdef to use for it? I'm happy to block it out if you want. I can also add a `swap()` method and default ctor so people on older compilers can at least do a manual move it they need to.

I thought it was best to omit the `unique_ptr_t` typedef since it doesn't really behave like a `std::unique_ptr<T>` and might be confusing for newcomers trying to use it as such.

llchan replied:
I just checked, and the interwebs seems to be telling me that rvalue refs were added in VS2010/MSVC16 and GCC 4.3. Since this a subset of the C++11 preprocessor check at the top of the file, I think we can leave the move ctor as is.

aardappel replied:
Ok, sounds good.

llchan commented on include/flatbuffers/flatbuffers.h:
```diff
@@ -542,57 +547,186 @@ struct String : public Vector<char> {
-  virtual ~simple_allocator() {}
-  virtual uint8_t *allocate(size_t size) const { return new uint8_t[size]; }
-  virtual void deallocate(uint8_t *p) const { delete[] p; }
+  FlatBuffer(Allocator *allocator,
```
Got it. Updated my clang-format rules too.

aardappel replied:
The Google C++ Style Guide has a built-in preset in clang-format :)

aardappel commented:
Ok, this now seems good in its current form, any further issues can be addressed in follow-ups. Thanks @llchan !

llchan commented:
Great! :)

Do you intend to bump the major version number, or is this fringe enough to stay on 1.x.x? It isn't really backwards compatible, so technically it should be a 2.x.x release. If we are going that route, I'd say let's hold off a bit and get the breaking gRPC changes in there as well.

aardappel commented:
There have been breaking changes before between subsequent 1.x.0 releases. So This would be in 1.7.0. I agree it would be good to do that release only once the dust has settled around the GRPC work too.
